test/gurka/test_loops.cc

```diff
@@ -0,0 +1,118 @@
+#include "gurka.h"
+#include <gtest/gtest.h>
+
+using namespace valhalla;
+
+void check_opposing(const gurka::map& map) {
+  // then we test the invariant that any edge should be the opposing edge of its opposing edge
+  // edge == opposing(opposing(edge))
```

Comment:
I'm wondering if we should also include this assertion somewhere in data processing to catch bad data in development, what do you think?

Reply:
we kind of do that already! its here: https://github.com/valhalla/valhalla/blob/master/src/mjolnir/graphvalidator.cc#L164-L179

it basically says if im assigning an opp index to an edge that already got one that means there were duplicates

Reply:
But that's only logging a debug statement to the console right? If violating the invariant `edge == opposing(opposing(edge))` leads to invalid memory access when querying the service, it seems like we should consider that a fatal error at data processing time.